daft/udf.py

```diff
@@ -567,3 +567,70 @@ def _udf(f: UserDefinedPyFuncLike) -> UDF:
         return udf
 
     return _udf
+
+
+class _DaftFunc:
+    """`@daft.func` Decorator to convert a Python function into a `UDF`.
+
+    Unlike `@daft.udf(...)`, `@daft.func` operates on single values instead of batches.
+
+    Examples:
+    >>> import daft
+    >>>:from daft import col
+    >>> @daft.func()
+    ... # or you can specify the return type like our existing daft.udf
+    ... # @daft.func(return_dtype=daft.DataType.int64())
+    ... def my_sum(a: int, b: int) -> int:
+    ...     return a + b
+    >>>
+    >>> df = daft.from_pydict({"x": [1, 2, 3], "y": [4, 5, 6]})
+    >>> df.select(my_sum(col("x"), col("y"))).collect()
+    ╭───────╮
+    │ x     │
+    │ ---   │
+    │ Int64 │
+    ╞═══════╡
+    │ 5     │
+    ├╌╌╌╌╌╌╌┤
+    │ 7     │
+    ├╌╌╌╌╌╌╌┤
+    │ 9     │
+    ╰───────╯
+    <BLANKLINE>
+    (Showing first 3 of 3 rows)
+    """
+
+    batch = udf
+
+    def __init__(self, func: UserDefinedPyFuncLike) -> None:
+        self._func = func
+
+    def __call__(self, *, return_dtype: DataTypeLike | None = None) -> UDF:
```

Comment:
I don't think this works because `__init__` will be called with the arguments `...` in `@daft.func(...)` whereas `__call__` will be called with the function being decorated, instead of the other way around